packages/compass-aggregations/src/modules/pipeline-builder/pipeline-mode.ts

```diff
@@ -26,9 +34,28 @@ const reducer: Reducer<State> = (state = INITIAL_STATE, action) => {
   return state;
 };
 
-export const changePipelineMode = (mode: PipelineMode): PipelineModeToggledAction => ({
-  type: ActionTypes.PipelineModeToggled,
-  mode,
-});
+export const changePipelineMode = (
+  newMode: PipelineMode
+): PipelineBuilderThunkAction<void, PipelineModeToggledAction> => {
+  return (dispatch, getState, { pipelineBuilder }) => {
+    // Sync the PipelineBuilder
+    if (newMode === 'as-text') {
+      pipelineBuilder.stagesToSource();
+    } else {
+      pipelineBuilder.sourceToStages();
+    }
+
+    dispatch({
+      type: ActionTypes.PipelineModeToggled,
+      mode: newMode,
+      stageOperators: getStageOperatorsFromPipelineSource(getState(), pipelineBuilder),
```

Comment:
Redux docs recommend to leave deriving state from actions to reducers, so in this case I would suggest to instead of doing this "fallback to current state" inside the `getStageOperatorsFromPipelineSource` method, just return the stageOperators from the builder here directly:

```suggestion
      pipeline: pipelineBuilder.pipeline,
```

and handle this in the reducer of the text-editor slice directly with something like

```js
return {
  ... other state
  stageOperators: action.pipeline ? getOperators(action.pipeline) : state.stageOperators
}
```